### DISCUSSION TITLE
[4.6] OCI - Build plugins jobs failing at npm install

### DISCUSSION THREAD

**Discussion by ddevadat** [Sunbird-Ed/Community#58] [Installation] [answered]
env details: sunbird 4.6 on OCI 
The below build job are failing 
 Build/Plugins/CollectionEditor
 Build/Plugins/ContentEditor
 Build/Plugins/GenericEditor

e.g when running Build/Plugins/CollectionEditor with release-3.6.0_RC1 getting the below error

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/postcss-cli/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/watchpack/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/watchpack-chokidar2/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN postcss-cli@8.2.0 requires a peer of postcss@^8.0.0 but none was installed.
npm WARN compression-webpack-plugin@10.0.0 requires a peer of webpack@^5.1.0 but none was installed.
npm WARN bower-resolve-webpack-plugin@1.0.5 requires a peer of enhanced-resolve@^3.1.0 but none was installed.
npm WARN filemanager-webpack-plugin@7.0.0 requires a peer of webpack@^5.0.0 but none was installed.
npm WARN fontmin-webpack@4.0.0 requires a peer of webpack@5.x but none was installed.
npm WARN ajv-keywords@5.1.0 requires a peer of ajv@^8.8.2 but none was installed.
npm WARN ajv-formats@2.1.1 requires a peer of ajv@^8.0.0 but none was installed.
npm WARN update-browserslist-db@1.0.5 requires a peer of browserslist@>= 4.21.0 but none was installed.
npm WARN postcss-load-config@3.1.4 requires a peer of postcss@>=8.0.9 but none was installed.
npm WARN postcss-load-config@3.1.4 requires a peer of ts-node@>=9.0.0 but none was installed.
npm WARN postcss-reporter@7.0.5 requires a peer of postcss@^8.1.0 but none was installed.
npm ERR! Linux 5.4.0-1066-oracle
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install"
npm ERR! node v8.10.0
npm ERR! npm  v3.10.10
npm ERR! path /var/lib/jenkins/workspace/Build/Plugins/CollectionEditor/node_modules/.staging/@types/node-71551420/package.json
npm ERR! code ENOTDIR
npm ERR! errno -20
npm ERR! syscall open

npm ERR! ENOTDIR: not a directory, open '/var/lib/jenkins/workspace/Build/Plugins/CollectionEditor/node_modules/.staging/@types/node-71551420/package.json'
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /var/lib/jenkins/workspace/Build/Plugins/CollectionEditor/npm-debug.log
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 236
Finished: FAILURE

**Comment by sharathkashyap** [comment]
@ddevadat I see that your node version is quite old
`node v8.10.0
npm ERR! npm v3.10.10` 
Could you please retry the steps with recommended version of node (i.e 12 or 14) and also npm being upgraded to (version 5)

**Comment by ddevadat** [comment]
The npm and node version are below

```
 /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm -version
6.13.4
/var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/node --version
v12.16.1
```

however the default version is

```
/usr/bin/node --version
v8.10.0

/usr/bin/npm --version
3.10.10
```

Somehow the jenkins build job is using the default ones. How do i change this behaviour

**Reply by sharathkashyap** [reply]
In Manage Jenkins Menu, Navigate to Global Tool Configuration option. 
**Manage Jenkins > Global Tool Configuration > NodeJS** 
 Check for NodeJS version installed. Change the version to recommended one.

**Comment by ddevadat** [comment]
There is no nodejs installed in the above navigation in jenkins.

Note that these jobs have successfully executed in the past. Its only failing when trying to run the build job again. Is there any clean up to be done if the jobs need to be run again

**Reply by sharathkashyap** [reply]
@ddevadat If in case the option doesn't exist. There is nodejs plugin available which allows the flexibility to have custom nodejs version associated with the specific job.

**Comment by ddevadat** [comment]
any update on this issue

**Comment by ddevadat** [comment]
For the job Build/Plugins/CollectionEditor, 
i tried editing the jenkinsfile, by replacing npm with /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm
e.g

```

              stage('Build') {
                    sh """
                        export version_number=${branch_name}
                        export build_number=${commit_hash}
                        rm -rf collection-editor
                        node -v
                        npm -v                        
                        #npm install
                        /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm install
                        cd app
                        bower cache clean
                        bower install --force
                        cd ..
                        #gulp build
                        gulp packageCorePlugins
                        
                        #npm run collection-plugins
                        #npm run build
                        #npm run test
                        
                        /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm run collection-plugins
                        /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm run build
                        /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm run test
                        
                        #cp collection-editor.zip collection-editor.zip:${artifact_version}
                    """
                }

```

its now failing at /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm run test with the below error

```

+ /var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm run test

> @project-sunbird/collection-editor@0.0.3 test /var/lib/jenkins/workspace/Build/Plugins/CollectionEditor
> karma start karmaconf.js


START:
29 07 2022 10:00:01.117:WARN [watcher]: All files matched by "/var/lib/jenkins/workspace/Build/Plugins/CollectionEditor/app/scripts/collectioneditor/collectioneditor-base-plugin.js" were excluded.
29 07 2022 10:00:01.119:WARN [watcher]: All files matched by "/var/lib/jenkins/workspace/Build/Plugins/CollectionEditor/app/scripts/collectioneditor/md5.js" were excluded.
29 07 2022 10:00:01.354:INFO [karma]: Karma v1.3.0 server started at http://localhost:9876/
29 07 2022 10:00:01.355:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
29 07 2022 10:00:01.360:INFO [launcher]: Starting browser PhantomJS
29 07 2022 10:00:01.492:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket /#lkxdCOLh3nZp1CQHAAAA with id 70931327
PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  SyntaxError: Functions cannot be declared in a nested block in strict mode
  at app/bower_components/contenteditor/index.js:1

Finished in 0.146 secs / 0 secs @ 10:00:01 GMT+0530 (IST)

SUMMARY:
✔ 0 tests completed
PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  TypeError: undefined is not an object (evaluating 'window.org.ekstep')
  at app/scripts/collectioneditor/bootstrap-editor.js:9

x Error while running the tests! Exit code: 1

PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  ReferenceError: Can't find variable: org
  at app/scripts/collectioneditor/collectioneditor-config.js:9

x Error while running the tests! Exit code: 1

PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  TypeError: undefined is not an object (evaluating 'window.org.ekstep')
  at app/scripts/collectioneditor/collection-editor-api.js:9

x Error while running the tests! Exit code: 1

PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  ReferenceError: Can't find variable: org
  at app/scripts/collectioneditor/manager/container-manager.js:9

x Error while running the tests! Exit code: 1

PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  ReferenceError: Can't find variable: org
  at app/scripts/collectioneditor/manager/metapage-manager.js:9

x Error while running the tests! Exit code: 1

PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  ReferenceError: Can't find variable: org
  at app/scripts/collectioneditor/service/collection-service.js:9

x Error while running the tests! Exit code: 1

PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  ReferenceError: Can't find variable: org
  at test/baseSpec.js:6

x Error while running the tests! Exit code: 1

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @project-sunbird/collection-editor@0.0.3 test: `karma start karmaconf.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @project-sunbird/collection-editor@0.0.3 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /var/lib/jenkins/.npm/_logs/2022-07-29T04_30_01_660Z-debug.log
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 1
Finished: FAILURE

```

**Comment by ddevadat** [comment]
is there any workaround or fix for this issue.
is it ok to comment out the test part while running the job
```
e.g
#npm run test
```

**Reply by sharathkashyap** [reply]
@ddevadat It should be okay to comment the test case run

**Comment by ddevadat** [comment]
any update on this issue

**Reply by sharathkashyap** [reply]
@ddevadat Sorry For the delayed response. Can you try commenting out test run. and retry the build.

**Comment by ddevadat** [comment]
heres what i tried 
1) use the complete path of npm in the jenkins definition
```
/var/lib/jenkins/.nvm/versions/node/v12.16.1/bin/npm
```

2) disable npm test
```
#npm run test
```

After doing this Build/Plugins/CollectionEditor -- successfull
But Build/Plugins/GenericEditor is still failing

```

ERROR: [Cobertura] No coverage results were found using the pattern 'coverage/PhantomJS*/cobertura-coverage.xml' relative to '/var/lib/jenkins/workspace/Build/Plugins/GenericEditor'.  Did you enter a pattern relative to the correct directory?  Did you generate the XML report(s) for Cobertura?
Finished: FAILURE

```
The job Build/Plugins/GenericEditor is referring
tag: release-3.6.0_RC1
repo: https://github.com/project-sunbird/sunbird-generic-editor.git

**Comment by luckynvdu** [comment]
@sharathkashyap Can we look into this and work with the reported to mark this closed?

**Comment by sharathkashyap** [comment, marked as answer]
Hello @ddevadat ,
            There has been couple of issues reported in other threads as well on similar grounds. Kindly have a look if this helps.
https://github.com/orgs/Sunbird-Ed/discussions/88#discussioncomment-3453311

**Reply by gandham-santhosh** [reply]
@ddevadat please let us know if any other information is required here or else please mark the thread
